External/FreeSASA/RDFreeSASA.cpp

```diff
@@ -0,0 +1,209 @@
+          BOOST_LOG(rdWarningLog) << "Atom " << atom->getIdx()
+                                  << " could not be classified" << std::endl;
+          success = false;
+        } else {
+          classification = classifier->class2str(cls, classifier);
+        }
+      }
+    }
+
+    radii.push_back(radius);
+    atom->setProp(common_properties::Atom::SASAClass, cls);
+    atom->setProp(common_properties::Atom::SASAClassName, classification);
+  }
+
+  return success;
+}
+
+double calcSASA(const ROMol &mol, const std::vector<double> &radii,
+                const SASAOpts &opts) {
+  return calcSASA(mol, radii, -1, opts);
+}
+
+double calcSASA(const ROMol &mol, const std::vector<double> &radii, int confIdx,
+                const SASAOpts &opts) {
+  PRECONDITION(mol.getNumConformers(), "No conformers in molecule");
+  PRECONDITION(confIdx < rdcast<int>(mol.getNumConformers()),
+               "Conformer index out of range");
+  PRECONDITION(mol.getNumAtoms(), "Empty molecule");
+
+  freesasa_parameters params = freesasa_default_parameters;
+  params.n_threads = 1;
+  switch (opts.d_alg) {
+    case SASAOpts::LeeRichards:
+      params.alg = FREESASA_LEE_RICHARDS;
+      break;
+    case SASAOpts::ShrakeRupley:
+      params.alg = FREESASA_SHRAKE_RUPLEY;
+      break;
+    default:
+      throw ValueErrorException("Unknown freesasa algorithm");
+  }
+
+  // sneaky, but legal :)
+  std::vector<double> coords(mol.getNumAtoms() * 3);
+  const RDGeom::POINT3D_VECT &vec = mol.getConformer(confIdx).getPositions();
+  for (size_t i = 0; i < mol.getNumAtoms(); ++i) {
+    coords[i * 3] = vec[i].x;
+    coords[i * 3 + 1] = vec[i].y;
+    coords[i * 3 + 2] = vec[i].z;
+  }
+
+  freesasa_result *res =
+      freesasa_calc_coord(&coords[0], &radii[0], mol.getNumAtoms(), &params);
+  if (!res) return 0.0;
+  CHECK_INVARIANT(res->n_atoms == rdcast<int>(mol.getNumAtoms()),
+                  "freesasa didn't return the correct number of atoms");
+
+  double sasa = res->total;
+  mol.setProp(common_properties::Molecule::SASA, sasa);
+  size_t i = 0;
+  for (ROMol::ConstAtomIterator at = mol.beginAtoms(); at != mol.endAtoms();
+       ++at, ++i) {
+    (*at)->setProp(common_properties::Atom::SASA, res->sasa[i]);
+  }
+
+  freesasa_result_free(res);
+  return sasa;
+}
+
+double calcSASA(const RDKit::ROMol &mol, const std::vector<double> &radii,
+                const RDKit::QueryAtom *query,
+                const SASAOpts &opts) {
+  return calcSASA(mol, radii, query, -1, opts);
+}
+
+double calcSASA(const RDKit::ROMol &mol, const std::vector<double> &radii,
+                const RDKit::QueryAtom *query,
+                int confIdx,
+                const SASAOpts &opts) {
+  calcSASA(mol, radii, confIdx, opts);
+  double total = 0.0f;
+  for (ROMol::ConstQueryAtomIterator at = mol.beginQueryAtoms(query);
+       at != mol.endQueryAtoms(); ++at) {
+    const Atom *atom = *at;
+    total += atom->getProp<double>("SASA");
```

Comment:
Harping on the same point about monomer info: this is going to throw a `KeyError` if the atom that matches doesn't have monomer info